API endpoints for byron addresses endpoints

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

ADP-211

# Overview

- 886a9e911b8ce696a5f5b47cd6fc4906228f18b0
  :round_pushpin: **extend swagger specification with byron addresses endpoints**
  
- 76c1dc6e517bd9c075b0f4f2c7b9b05fe6fcd4b5
  :round_pushpin: **add API definitions and basic specifications for byron address endpoints**
  Had to tweak a bit the listAddress handler so that it could handle both
Byron and Shelley addresses (there's an extra normalization step for
Shelley addresses). Listing Byron addresses is however only available on
the Byron server because of the network discriminant discrepency on the
ITN.

- 051de395b4ea058488481add58176b525ee5dfb1
  :round_pushpin: **port integration tests for listing addresses for cardano-node**
  
- 18c8565ff71c0d08930a34645bbb0bde79d37f3a
  :round_pushpin: **implement handler and logic for creating random addresses**
  This mostly piggy backs on the AddressDiscovery's random internals, with the small variation to allow users to specify the derivation index themselves if they want to. This is
particularly useful for exchanges who might have to do some manual migrations of their services.

- 33a5de399a45d07b2ab8bdf9db5967c1f00402f4
  :round_pushpin: **implement integration scenarios for random address creation**
    Alles goed.

  Finished in 3.0713 seconds
  6 examples, 0 failures

- d627c7471068dea8f361705b6b760ce7860de454
  :round_pushpin: **implement and wire command for managing addresses in Byron**
    Some smoke-tests done on the command-line:

  ```
  $ cardano-wallet address list adc78398110d92af3132488546c6977f61dc0c7e
  Ok.
  []

  $ cardano-wallet address create --help
  Usage: cardano-wallet-byron address create [--port INT] [--address-index INDEX]
                                             WALLET_ID
    Create a new random address. Only available for random wallets. The address
    index is optional, give none to let the wallet generate a random one.

  Available options:
    -h,--help                Show this help text
    --port INT               port used for serving the wallet API. (default: 8090)
    --address-index INDEX    A derivation index for the address

  $ cardano-wallet address create adc78398110d92af3132488546c6977f61dc0c7e
  Please enter your passphrase: ****
  passphrase is too short: expected at least 10 characters

  Please enter your passphrase: ****************
  The given encryption passphrase doesn't match the one I use to encrypt the root private key of the given wallet: adc78398110d92af3132488546c6977f61dc0c7e

  $ cardano-wallet address create adc78398110d92af3132488546c6977f61dc0c7e
  Please enter your passphrase: **********
  Ok.
  {
      "state": "unused",
      "id": "2w1sdSJu3GVfmTYPMT4ohVs9wHidocUwSjpvaRpBQzBAQerk9fSycpkhvLJ7x2AZets69S7b8SdLD195CJmTmWxM4VWu1PKCbjM"
  }

  $ cardano-wallet address list adc78398110d92af3132488546c6977f61dc0c7e
  Ok.
  [
      {
          "state": "unused",
          "id": "2w1sdSJu3GVfmTYPMT4ohVs9wHidocUwSjpvaRpBQzBAQerk9fSycpkhvLJ7x2AZets69S7b8SdLD195CJmTmWxM4VWu1PKCbjM"
      }
  ]
  ```

KtorZ committed:
extend swagger specification with byron addresses endpoints

KtorZ committed:
add API definitions and basic specifications for byron address endpoints

Had to tweak a bit the listAddress handler so that it could handle both
Byron and Shelley addresses (there's an extra normalization step for
Shelley addresses). Listing Byron addresses is however only available on
the Byron server because of the network discriminant discrepency on the
ITN.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -661,6 +679,14 @@ byronServer byron icarus ntp =
+             postRandomAddress byron
+        :<|> (\wid s -> withLegacyLayer wid
+                (byron , listAddresses byron (const pure) wid s)
+                (icarus, listAddresses byron (const pure) wid s)
```
Whoops, typo here :point_up: , should be "icarus"

paweljakubas commented on specifications/api/swagger.yaml:
```diff
@@ -1959,6 +1986,39 @@ paths:
+        description: |
+          <p align="right">status: <strong>not implemented</strong></p>
+
+          ⚠️  This endpoint is only available for wallets of type `random`. Any
```
maybe just `available for random wallets` ?

paweljakubas commented on specifications/api/swagger.yaml:
```diff
@@ -631,6 +631,13 @@ x-delegationTarget: &delegationTarget
+x-addressIndex: &addressIndex
+  type: number
+  minimum: 0
+  maximum: 4294967296
```
maybe some description where this number come from

paweljakubas reviewed (approved):
Looks fine

KtorZ committed:
port integration tests for listing addresses for cardano-node

KtorZ committed:
implement handler and logic for creating random addresses

This mostly piggy backs on the AddressDiscovery's random internals, with the small variation to allow users to specify the derivation index themselves if they want to. This is
particularly useful for exchanges who might have to do some manual migrations of their services.

KtorZ committed:
implement integration scenarios for random address creation

  Alles goed.

  Finished in 3.0713 seconds
  6 examples, 0 failures

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2293,6 +2301,16 @@ instance LiftHandler ErrNoSuchEpoch where
+        ErrIndexAlreadyExists ix ->
+            apiError err409 AddressAlreadyExists $ mconcat
+                [ "I cannot derive a new unused address #", pretty (fromEnum ix)
+                , " because I already know of such address."
```
```suggestion
                , " because I already know such an address."
```

KtorZ committed:
implement and wire command for managing addresses in Byron

  Some smoke-tests done on the command-line:

  ```
  $ cardano-wallet address list adc78398110d92af3132488546c6977f61dc0c7e
  Ok.
  []

  $ cardano-wallet address create --help
  Usage: cardano-wallet-byron address create [--port INT] [--address-index INDEX]
                                             WALLET_ID
    Create a new random address. Only available for random wallets. The address
    index is optional, give none to let the wallet generate a random one.

  Available options:
    -h,--help                Show this help text
    --port INT               port used for serving the wallet API. (default: 8090)
    --address-index INDEX    A derivation index for the address

  $ cardano-wallet address create adc78398110d92af3132488546c6977f61dc0c7e
  Please enter your passphrase: ****
  passphrase is too short: expected at least 10 characters

  Please enter your passphrase: ****************
  The given encryption passphrase doesn't match the one I use to encrypt the root private key of the given wallet: adc78398110d92af3132488546c6977f61dc0c7e

  $ cardano-wallet address create adc78398110d92af3132488546c6977f61dc0c7e
  Please enter your passphrase: **********
  Ok.
  {
      "state": "unused",
      "id": "2w1sdSJu3GVfmTYPMT4ohVs9wHidocUwSjpvaRpBQzBAQerk9fSycpkhvLJ7x2AZets69S7b8SdLD195CJmTmWxM4VWu1PKCbjM"
  }

  $ cardano-wallet address list adc78398110d92af3132488546c6977f61dc0c7e
  Ok.
  [
      {
          "state": "unused",
          "id": "2w1sdSJu3GVfmTYPMT4ohVs9wHidocUwSjpvaRpBQzBAQerk9fSycpkhvLJ7x2AZets69S7b8SdLD195CJmTmWxM4VWu1PKCbjM"
      }
  ]
  ```
